Citation dialog: allow to insert annotations

abaevbog opened the pull request:
- update `EditorInstance` to allow creation of a note from annotations without saving the item to DB. When the image or ink annotations are being inserted, they cannot be added into the note via attachmentKey (since it would require to save entries to database). Instead, they are embedded as a src="data:image/..." string.
- `SearchHandler` will match annotations based on their content, as well as the top-level item. Same applies to notes, for consistency. The first two commits are minor search tweaks to allow that.
- returned child annotations will appear as child rows in list mode. Child rows can be hidden by collapsing the top-level item.
- in both list and library mode, one can select multiple annotations, which will all be added when the dialog is accepted. One can also select an attachment or top-level item, in which case all annotations that the item contains will be added.
- when displaying selected/open/cited items, only keep ones that have annotations

<img width="804" alt="Screenshot 2025-06-24 at 9 01 43 PM" src="https://github.com/user-attachments/assets/3ba36ed3-8639-4371-8eef-442bdd0ac26a" />
<img width="799" alt="Screenshot 2025-06-24 at 8 56 44 PM" src="https://github.com/user-attachments/assets/68ae0739-c2e1-46fd-90d9-9d72ea6fb8bf" />

Fixes: #5314

abaevbog commented:
For now, I hard-coded the "Add note" button to open in "Add annotation" mode, for testing. 

@yexingsha, how does this look to you?

As discussed before in https://github.com/zotero/zotero/issues/5314#issuecomment-2978984429, we may want to add "Add annotation" as another button, next to "Add note". @adomasven what would be a good starting point for this?

adomasven commented:
Latest build of connector (and whenever a new Connector beta is released) should display an "Add annotation..." menu option. I've added that since we don't have icons in the Google Docs UI. For others ideally I wouldn't have to make my own graphics.

yexingsha commented:
> One can also select an attachment or top-level item, in which case all annotations that the item contains will be added.

I'm a little wary of this, because it seems very easy in list mode to click on a parent item trying to collapse it, only to add all its annotations (which in a book can be a lot) by mistake. I'm wondering if "add every annotation" is a common enough use case to justify it being the main action when you click on a parent item. 

I can't get the "Add note" button to work as intended, so the following UI stuff is just from the screenshots:

- The hint text should say "Search for annotations".
- The numbers after the "Selected Items" and "Open Documents" don't make much sense for annotations. I was thinking we either say something like "x Annotations" or just get rid of the numbers altogether.
- Again, not sure if we should still keep the "add all" for annotations in selected items. Either way, expanding this section by default is probably the correct choice. 
- In Library Mode, the two Open Documents cards seem to be missing... something? Is it because they are image or ink annotations? In that case, they should still keep to the format: first line (small, gray) is title, second line is "Image/Ink annotation", and the third line is empty.
- In List Mode, the second line of parent items is misaligned from the first.
- In List Mode, the annotations should have 16px inset, instead of the 10px right now.

I have made a set of SVG icons for word processor plugins, which would look like this in Word:

<img width="700" alt="ribbon" src="https://github.com/user-attachments/assets/701f5b37-68c8-4914-a419-886453b7495c" />

[plugin-icons.zip](https://github.com/user-attachments/files/20916922/plugin-icons.zip)

All icons are in light and dark mode, with sizes at 16px, 26px, and 32px. @adomasven let me know if you need more icons, sizes, etc.

abaevbog commented:
> I'm a little wary of this, because it seems very easy in list mode to click on a parent item trying to collapse it, only to add all its annotations (which in a book can be a lot) by mistake. I'm wondering if "add every annotation" is a common enough use case to justify it being the main action when you click on a parent item.

Fair question. Initially this was brought up in https://github.com/zotero/zotero/issues/5314#issuecomment-2978984429. I guess the main idea is that this dialog is essentially a replacement of the current workflow: select the item(s) in the library tab, click "Add note from annotations" in the context menu, and then insert that note. So I suppose if you can create a note from the attachment and the top-level item, it should still be possible here. Should we make the twisty bigger (or maybe not twisty itself but rather the clickable area around it), to make mis-clicks less likely?

> The numbers after the "Selected Items" and "Open Documents" don't make much sense for annotations. I was thinking we either say something like "x Annotations" or just get rid of the numbers altogether.

If we do keep the regular items as options in "Open Documents", "Selected Items", and etc., we can totally hide the counts if that sends conflicting info. I see your point that showing how many items are selected or opened isn't really that useful here. Instead, maybe it would be helpful to display how many annotations each item has? On a similar note, do you find the item cards in library mode a bit hard to read? Annotation cards have the top-level title at the top, which is not super indicative of the fact that this is an annotation vs a top-level item. Maybe it's worth to add the item icon somewhere to make it clearer it it's the annotation or the entire item selected? Or maybe I'm overthinking it?

> Again, not sure if we should still keep the "add all" for annotations in selected items. Either way, expanding this section by default is probably the correct choice.

Right, this goes back to the first question you brought up. But here I do see one good use case. One may go through many annotations (in a saved search, maybe), select them all, and then open the "Add annotations" dialog intending to insert them all. A one-click "Add all" can save a few clicks here.

> In Library Mode, the two Open Documents cards seem to be missing... something? Is it because they are image or ink annotations?

Opened documents at the moment are, well, just the normal opened top-level documents (that do have some annotations). Clicking one card adds all of the annotations in that document. Since annotation cards have text, comment, and the title of the top-level item, they have 3 lines. Opened documents only have two: the title and description, just like when citations are being added. The same would happen with selected items if top-level items are selected (in the screenshot, both items just happened to be annotations). Differentiating between cards of top-level items and annotations is somewhat hard, so maybe adding icons would help? Or we could consider pulling out all annotations from opened documents and displaying all annotation card, instead of just one card per top level item? I'm afraid this could get overwhelming though.

abaevbog commented:
> I guess we don't need to make top-level items or attachments selectable if we're concerned about accidental adds. You can always just select all the annotations under an item. That's a lot more annoying for multiple items, though.

Sounds good, let's try that. But that is a problem just in list mode, so do we want to still allow to select top-level items or attachments in itemTree in library mode? No reason not to, besides inconsistency. We could add a small button to top-level items in list mode that would select all of its child annotations, without accepting the dialog? So you don't have to CMD-click every single annotation.

> And looking at the screenshot, maybe it's too confusing. It's not totally clear what will happen if you select an item or attachment. So I guess we can limit this to annotation items?

Currently, if a top-level item or attachment with annotations is selected, it appears the same way as when a citation is being added, and clicking on it will add all of its annotations. It may make it easier if we only display annotation items, that's true. Because right now there are essentially two types of item cards (one for one annotation and one for an annotation ancestor), and you need to distinguish between the two. What about cited items and open documents? Should we just not include them?

> And I think we can hide non-matching sibling items (e.g., notes or attachments without annotations) altogether.

You mean in itemTree, in library mode? Good point! Notes and attachments without annotations are 100% not useful here.

yexingsha commented:
It works, thank you!

I guess the easiest thing to do would be to make parent items and attachments non-selectable. That way, clicking on a parent item always expands/collapses it, so we don't even need a visible click target for the twisty. Though this also raises questions about what to do with them in keyboard navigation. Probably focus ring instead of the focus background color?

And now I see what's going on with Selected Items and Open Documents. I agree that showing all annotations under Selected and Open items is too much, but still, being able to quickly select annotations from the items you have selected/open would be useful. I think we can attempt a middle ground:

- If an item is selected at the parent item or attachment level, or if a document is open, show the top level parent item, with option to expand and show all its annotations;
- If an item/attachment AND one or more of its annotations are selected, show the parent item expanded, but with only the annotation(s) selected visible, the rest hidden behind a "X more..." button;
- If an annotation is directly selected, do what we do now.

If we adopt this approach, then we probably need to a tooltip for "Add all" that explains it will only add all the directly selected annotations.

This does get trickier in Library Mode. I don't think it really makes sense to expand all the annotations under a parent item in card format, but again, adding all the annotations under one item is probably not that common. So what if clicking a Suggested/Open parent item sends you to the collection it's in, puts it at the top of the view, and expands it to show all its annotations?

Mocks below. Other changes:

- Probably need a setting to toggle comments on/off
- In List Mode:
	- Added buttons to select all child annotations (slightly tweaked from the button from Attachment section in itemPane)
	- Adjusted twisty padding so that parent items line up with Selected Items header
- In Library Mode:
	- Reformatted cards. Parent items get icon at first line, with an additional line "X Annotations..."; annotations get icon at second line

<img width="1000" alt="annotation-list" src="https://github.com/user-attachments/assets/2e9222eb-09c7-4ab7-9fd7-8a8200d591ce" />

<img width="1000" alt="annotation-library" src="https://github.com/user-attachments/assets/b14c5aa5-b006-4622-a873-912d44ab640e" />

abaevbog commented:
> And I think we can hide non-matching sibling items (e.g., notes or attachments without annotations) altogether.

I realized we already have https://github.com/zotero/zotero/pull/5315 PR to hide context rows. In that PR, context rows are hidden based on a pref, so we would just always enforce it here.

Should I include those commits into this PR as well, or just wait for it to be merged?

abaevbog commented:
As another note, now that we want to be able to display all annotations of an opened/selected item in list mode, we probably have to display list nodes in a windowed list, since there can be a lot of annotations even in just one item. Currently, we avoided overloading DOM by just rendering at most `ITEM_LIST_MAX_ITEMS` nodes (currently 50) - but that simple approach won't really work anymore.

yexingsha commented:
Re: pseudo collections, it feels awkward that Add Annotation is the only dialog that has to resort to this. Also, when people use this dialog, it's likely they won't have any annotation selected, so the suggestion area will be blank most of the time. I guess we can write something like "Selected Items and Open Documents are displayed as collections"?

But this is probably the least bad option. If we do this, then we don't need to put icons in the cards either, so that we can display a bit more of the annotation and comment text.

adomasven commented:
I don't think we necessarily need to hang up on what we do with item and note citation for this dialog. Perhaps we should only keep the library mode, and remove the suggested area completely, keeping only the collection and item trees.

Before the new citation dialog, when I used to plan what the reworked dialog would look like, I imagined a version of the item tree that has collapsing sections, kind of like what Windows Explorer does in this screenshot I found online.

![image](https://github.com/user-attachments/assets/bde45f03-772c-4659-b866-2bf3b46ef356)

We could have "Selected items" collapsible section at the top of the item tree, followed by "All items". I think it's better than a pseudo-collection for discoverability, although maybe somewhat awkward.

Just throwing around ideas, let me know what you think.

abaevbog commented:
@adomasven, on a separate note, now that we want to be able to show all annotations of a top-level item in list mode, I believe we have to handle displaying of arbitrary number of annotation nodes in the list (instead of just allowing at most 50 nodes, as done now). The nodes in the array potentially have different heights, so I don't think `VirtualizedTable` would work here (the way it works for collectionTree). Is `windowed-list.js` the best thing to use? With setting `customRowHeights` to indicate how tall each node is. I did get a minimal example working but wanted to double check because it doesn't look like `windowed-list` itself is used anywhere else besides `VirtualizedTable`.

adomasven commented:
> Is `windowed-list.js` the best thing to use? With setting `customRowHeights` to indicate how tall each node is. I did get a minimal example working but wanted to double check because it doesn't look like `windowed-list` itself is used anywhere else besides `VirtualizedTable`.

See https://github.com/zotero/zotero/pull/3460

This was implemented for custom row heights for spacers in the collection tree. `VirtualizedTable.updateCustomRowHeights()` can be used. VirtualizedTable has keyboard and mouse I/O handlers, so I would use that unless it limits you in some fundamental way, in which case you'd have to work with the raw WindowedList indeed.

adomasven commented:
> So part of me thinks to embrace these differences and not try to overcomplicate the library mode with selected/open items when list mode is probably just going to be better for that?

It's hard to claim something with high certainty without doing user testing, as it's unclear how much of an "advanced user feature" citing annotations is going to be, but for regular non-technical users figuring out that different modes are better for finding selected/library items IMO is just too much. I think generally the expectation is that the majority of users use either one or the other mode, and once they settle with one, they don't even think (forget) that another mode exists.

However for annotation citing if one mode is much superior to the other, or if one mode makes more sense than the other, then maybe we shouldn't try to make both work.

abaevbog commented:
@adomasven, I think the list mode rewrite to use `virtualized-table` has been going fairly well. I'm still doing some more testing and tweaking but I think the old behavior of citation dialog should be mainly replicated, together with the added new design for annotations. I wanted to see what you think about the list mode now?

A few minor notable changes are:
- I added custom heights to be set as actual heights of  `virtualized-table` rows, otherwise some rows would not get properly aligned.
- the section headers, including collapsible "Selected Items", are now all table rows. It raised the question on what to do with focusable button to expand/collapse the section and "Add all" button. I thought it's best to make them un-focusable and just let right/left arrows on the header row expand/collapse the section, while activating the row essentially clicks on the "Add all" button. It seemed like a reasonable compromise to me, especially because that way there is only 1 (instead of 3) Tabs between the bubble input and the items list.
- now there is way less common logic shared by the library and list mode, so a lot of it was moved from `IOManager` and abstract `Layout` class into the actual layouts. I also moved whatever was possible into `Helpers` to keep the main file less bloated. 
- as discussed earlier, the top-level items are selectable for keyboard navigation but clicking on them will only expand/collapse them to reveal annotation rows.
- I overrode some instances of mouse and keydown handling but the bulk of it is still handled by `virtualied-table` itself

With this, I am now circling back to the best solution for the selected/open items in library mode, since we never finalized it. For now, I excluded all non-annotation items from the suggested items area (because it was one of the earlier suggestions). I also added a sketch of the pseudo-collection above "My Library" to see what it looks like - seems quite useful to me, though it fully erases the point of having a more prominent horizontal suggested items area.

Two new ideas I had for suggested items:
- We can separate selected annotations from selected top-level items by placing them into two separate decks ("Selected annotations" and "Selected items"?). That would make it easier to indicate that the items cards are different, maybe allow one to easily add all annotations while ignoring non-annotation items, and etc. It still doesn't solve the issue of how to allow one to see not-selected annotations of a selected top-level item (or what clicking on the top-level item should do) but it could add some clarity.
- I thought more about your [last suggestion](https://github.com/zotero/zotero/pull/5356#issuecomment-3011840655), and my main concern was that displaying selected/open items at the top of `itemTree` would feel very strange when you switch between collections. But what if we just place selected/open items into a new (smaller) `itemTree` and place it into the suggested items area, instead of horizontal item cards, when annotations are being added? That would make it easy to interact with those items without conflicting with the main itemTree. Two tables may look a bit strange but functionally it would solve our issues. It's fairly similar to the "Selected/Open" pseudo collection I already added - but occupying a more prominent spot in the dialog. It would open a whole new set of questions (e.g. resizing, handling itemTree menubar for multiple itemTrees, etc.) but still I thought it's worth considering.

Let me know if you have any new thoughts on the library mode question!

(Also cc @yexingsha  on the library mode selected/open suggestions above)

abaevbog commented:
Thank you for the feedback @yexingsha!

> Pseudo collection of selected/open/cited items. After testing it, I feel like it makes the suggestion section mostly redundant; however, if we get rid of this section, we need to figure out a different way to add all selected annotations in library mode.

That is a very good point. I somewhat overlooked the question of adding all selected annotations with one click...

> An extra itemTree replacing the suggestion section. [...] But the vertical space will be really tight.

Yeah, I briefly tried this and it doesn't (really) fit into the existing layout at all. It ends up looking quite busy and the space is way too narrow. So it's something to keep in mind as a possibility but maybe not in the current layout of the citation dialog.

> Keep the suggestion area, but separate selected annotations from selected top-level items. I quite like this idea! 

Right, this way we keep the existing suggested area as a quick way to access most useful entries while addressing the possible confusion caused by different types of item cards mixed together. With this separation, I think when a suggested top-level item card is clicked, it would be just fine to move to that item in the library and (maybe) select all of its annotations. Then, if you do want to add just all annotations from that item, it's just a matter of an extra Enter keypress. But if you want to select some specific annotations, you also have them all expanded. So, unless there are any more ideas (or issues with this approach), I think we'll try that?

It _may_ turn out that seeing the annotations of selected top-level items together is handy. For instance, maybe you have item A and item B selected, and you want a few annotations from item A and a few from item B. They maybe in very different parts of the library, so clicking on a item A in suggested area can land you very far from item B. Not sure how much of a concern that is. But, if needed, we can always bring the pseudo-collection from point 1 as an addition to this. Then, clicking on item A would take you to that pseudo-collection, and item B would be next to it. Not sure if it's really necessary at this point but just an option for the future.

abaevbog commented:
@adomasven, thank you for your feedback! 

> When typing in the list mode, if I press up/down, I see the list being navigated, but pressing left/right to expand items does nothing. I have to press tab for left/right to start working. In Library mode focus from the input is removed and keyboard works as expected.

Right, I wanted to preserve the existing behavior of list mode in citation dialog, where arrow up/down changes the selected item but doesn't steal focus from the input. It is more reliable than sending focus into the list on the current item and then refocusing the input if one tries to type due to different locales and all sorts of edge cases with non-english keyboards. I don't think we can handle expanding and collapsing of selected container items with right/left arrows from the focused input (same way we handle arrow up and down now), since these arrows need to move the cursor between typed characters. So ... I suppose the question is what would be more important here - keeping the keyboard nav consistent with the citation dialog (keep focus in the input)  VS allowing arrow right/left to expand/collapse container items whenever they are selected (focusing item list on arrow up/down and refocus input when something that [seems like a printable character](https://github.com/zotero/zotero/blob/7beed24756bc19b4187ad8fc35b662ae3f628d0c/chrome/content/zotero/elements/bubbleInput.js#L563) is pressed). We could try to combine these - e.g. keep focus in the input and the first arrow right/left would perform the expand/collapse action on the selected container item (if there is one), while subsequent ones would move the cursor in the input. It is not standard but might be fairly intuitive.

> The fact that keyboard navigation gives only a faint outline to items, but mouse-hover gives a grey highlight is confusing. Why are they different?

We should probably tweak the styling here. The `:hover` effect is different from the `.selected` effect here, the same way it is different in the citation dialog. But it makes sense to have a special hover effect for container items that are clickable but not actually selectable.

> Pressing on the "All annotations" button expands the item (and also other items in the same section ??)

That is certainly not intended! I have not been able to run into this yet - could you share what the sections look like when this happens for you?

>  I think given we're using the same bubble input here, and you can add multiple annotations, we should bubbleize them like we do for citations, and use the same confirmation logic.

I see your point. We briefly discussed this in the linked issue a while back https://github.com/zotero/zotero/issues/5314#issuecomment-2978984429. Essentially, the idea was to have "Add note" interface but with multi-selection, since there is no reordering of annotations. But if we are running into the issue of accidentally accepting the dialog, I suppose there is no reason not to bubbleize annotations, and just disallow drag-drop reordering on them.

> When expanding and collapsing items in the list view, if there are only a few items being displayed, when the dialog resizes, there's a delayed pop-in of the items

I see what you mean. I'll see if we can maybe tell virtualized-table what height is will be before the window is actually resized.

>  I still think we should give collapsible sections in the main tree a try

For sure, let's give it a shot! I expect the approach will be similar to what we did with annotation rows. Each collapsed section will be a row with a special `renderItem` handling that will render a single `<span>` column with the "Selected"/"Open"/"Cited" label in the center with a twisty somewhere next to it. And, when the row is toggled to be expanded, we'll display relevant rows below them, maybe with an extra row in the end to serve as a clear separator. Let's try it!

yexingsha commented:
> All to say, the natural way to figure out which annotations to insert is the PDF viewer. This dialog can be functional, and better than nothing, but I don't think it's the right approach for citing annotations.

I don't disagree, and I think we will plug in a reader instance when we eventually put this plugin into a sidebar. But this gave me an idea. Instead of "Selected Items" only working on the itemTree in library tab, can we also include selected annotations in the active reader tab? This makes it much easier to browse the document in the main Zotero window, select the annotation(s) you need, and simply use this dialog to add the annotation(s) you have selected. This will probably be the main workflow to add annotations.

In addition, there are ways we can improve the search experience in this dialog as it is without plugging in a reader instance, which I agree will be too small to be useful here. In list mode, we can fully display the annotation text/image and comments, but I'm not sure if that makes it easier to search for the annotation you need. Maybe we should show two lines by default, but each individual annotation can be fully expanded:

<img width="1000" alt="annotation-list" src="https://github.com/user-attachments/assets/be007bd4-1343-48f6-9bb7-15389970e81f" />

- Right now I'm limiting image height to 100px.
- If the entire annotation can be fully displayed within 2 lines, then there is no twisty.

In Library mode, we can add the side bar that we have in main library view, with a button at the bottom to toggle it on/off:

<img width="1000" alt="annotation-library" src="https://github.com/user-attachments/assets/4ead21eb-8475-4972-a76f-226e04e64c6d" />

abaevbog commented:
> Given that we'll be bubbleizing the annotations, we will likely want to be able to preview them too.

Good point! That's something to think about - if we have the itemPane for the preview, it makes sense for a selected bubble preview to appear there. The question would be how to reconcile it with preview of whatever else could be selected in the itemTree at the same time. I'll think about what's the nicest way to do it.

> Also I saw your earlier comment that you cannot reorder annotations, but why would that be? 

The assumption was that annotations should be inserted in the order in which they appear in the document. Is that not the case? Should we sort the annotations initially when they are bubbleized but allow them to be rearranged, if needed?

adomasven commented:
> I actually wanted to do this way back when we first started redesigning the citation dialog. The main concern at the time was whether we could reliably bring the Zotero main window to the front, and then the word processor (or browser in the case of GDocs) to the front again, especially on windows and linux. @adomasven do you think that's possible?

Technically just focusing the Zotero window when another app is focused is often not possible. We get focus now by making Zotero open a new window (citation dialog), which most OSes put on top. In practice, if we really wanted to do this, we could explore hacky ways to make this work, like opening a temporary dialog, which focuses the Zotero citation interface. This could even maybe be the progress bar that we now show in most cases anyway. The risk would be that it might not work in some cases/for some configurations, so testing would be necessary.

I'd like to hear what @dstillman thinks about this (citing directly from the library window, by e.g. showing the bubble input at top/bottom). We've discussed the fact that Mendeley Desktop and Endnote allows this in the past (and users used to request it, although we haven't received requests recently), and it maybe nice to have, but as with all things, you have to balance accessibility, good UX and ease of use and discovery.

I assume if we went that way, we'd have to support all other citing dialogs too, because we'd definitely get many requests once people saw it was possible for annotations.

abaevbog commented:
I suppose I should hold off with the rest of the changes here until we decide if citing annotations will happen through the dialog or the main window, since in the latter case I assume the UI will be very different from what we have in the dialog so far.

adomasven commented:
> It makes sense, from the perspective that we already have all the interface necessary for navigating the reader, so it seems redundant to do a mini version of that in the citation dialog. At the same time, in my opinion, some aspects of the citation dialog make searching across different groups of items easier vs the main window. Even searching in the "Opened Documents" group to me seems easier in the dialog than in the main window (via Tabs menu). And if we load the cited items in the current document, we would need to display them in the main window as well somehow.

Right, I think this is a question of interface clarity vs duplicating what we already have. Technically citing and reading/annotating are two separate tasks which happen in a different mindspace, have different processes. In practice, the UI is similar enough that it's questionable whether it's worth duplicating it (and also whether there may be benefits to use existing UI, like annotating and citing in a single step).

There's also the tertiary question of performance and such, as loading the PDF reader isn't instantaneous even on new hardware, and many people are absolutely not on it. So if the annotation citing dialog needs to load the PDFs to be able to add annotations, that's perhaps non-ideal.

@yexingsha I'd like to hear your thoughts, and also, pinging @dstillman once more.

adomasven commented:
> * We essentially put the dialog into the main window as a `bubble-input` component somewhere, and you can accept from there once you are done.

The biggest question with this approach to me is how/where do we add the additional UI elements and keyboard interactions needed to add annotations into the bubble input. I like this in theory, but in practice the issue is that the main UI is quite noisy already. On the other hand, for an advanced user that's comfortable with the UI, this would probably be the preferred way to do all citing.

I guess another question to answer here, is whether citing annotations is an "advanced user feature", and we don't expect everyone to use it. Because if it were, we could just go for a more advanced/complicated UI choice and not spend too much time thinking on how to make it work for an inexperienced user.

On the other hand, if we want this to work for a less experienced user, a simpler dedicated interface is better for them.

abaevbog commented:
> The biggest question with this approach to me is how/where do we add the additional UI elements and keyboard interactions needed to add annotations into the bubble input.  I like this in theory, but in practice the issue is that the main UI is quite noisy already.

Right, that was my consideration behind the idea of "shortlisting" annotations from above. If we add a `bubble-input` in this way or another, essentially, where would it go? If it's some place very prominent, it will likely feel quite crowded. And if it's not, it will be hard to discover. Maybe it could be another tab in contextPane, similar to item notes? Then, one could drag-drop annotations from the sidebar on the left into the `bubble-input` on the right.

> I guess another question to answer here, is whether citing annotations is an "advanced user feature", and we don't expect everyone to use it.

I may be mistaken but I _think_ it can count as an advanced feature. I have only found a few posts on the forums asking about it:

https://forums.zotero.org/discussion/comment/490830/#Comment_490830
https://forums.zotero.org/discussion/100321/cite-pdf-annotation-from-markdown-editor-with-autocompletion
https://forums.zotero.org/discussion/104449/citing-annotation-comment
https://forums.zotero.org/discussion/96195/cite-single-annotations-with-word-processor

But then again, it's plausible that everyone is just used to the workaround of creating a note from annotations and inserting it.

yexingsha commented:
> Maybe it could be another tab in contextPane, similar to item notes? Then, one could drag-drop annotations from the sidebar on the left into the bubble-input on the right.

I was thinking it would go either under the toolbar, or at the bottom of the itemTree/reader area, but I like this better. It doesn't add more noise to the interface, and it can persist across library and reader tabs without shifting position.

> On the other hand, for an advanced user that's comfortable with the UI, this would probably be the preferred way to do all citing.

I agree, and I guess if we offer this as an "advanced" interface option alongside the citation dialog, it's less problematic if we can't get window focus to switch consistently.

> I guess another question to answer here, is whether citing annotations is an "advanced user feature", and we don't expect everyone to use it.

I don't think annotation citing is an advanced user feature; I think it's more about people's research field and what they are writing (e.g. a book review is much more likely to contain direct quotes than a scientific paper) than their experience with the software. At least one of the forum posts seems to be from a new user.

abaevbog commented:
> (e.g. a book review is much more likely to contain direct quotes than a scientific paper) than their experience with the software.

Yeah, good point!

So then, suppose we have a mini-`bubble-input` in the context pane of the reader. As one reads, they can drop annotations there. And then, if they press the "Add annotations" button in word/google-docs/etc., we immediately insert them?

And if one presses "Add annotations" button with nothing in the new mini-`bubble-input`, we use the current workflow with the dialog in library mode and etc. If one goes back to the main window, finds an annotation there, they can add it into the dialog via context menu (e.g. "Add to citation dialog") or drag-drop?

Is this roughly how this workflow would happen?

abaevbog commented:
> For people more comfortable with the dialog, they see the dialog, which contains the content in the bubble-input if that's available. They can accept that, or search for different annotations. If they can't find what they need easily, then they will click a "Open in Zotero" button, which opens the parent item in the main window and brings up the context pane Cite tab that contains the bubble-input.

Gotcha. In this case, does that mean that there would be two active instances of `bubble-input` at the same time from where one can accept what to insert? One in the dialog, and one in the main window. If that's the case, does that seem a bit confusing? Being able to confirm an action like which annotations to insert from multiple places seems unusual (and a bit confusing) to me...

> I think when people make annotations, they usually aren't looking to cite them immediately; they only decide which annotations to cite when they get down to writing.

As a reminder for myself, the main problem we are trying to address is how to allow the user to select annotations from the reader (where it is easier to navigate many annotations). We are also not expecting one to keep track what to cite during reading (which is fair, I think), and expect them to make a selection after opening an interface to cite annotations. (Right? Correct me on the above if that's off).

So what does a separate "Cite tab"  in the item/context pane add VS just a context menu "Add to citation dialog" on the citation in the sidebar/itemTree/etc? If one looks at the opened citation dialog and can't easily find exactly what they are looking for, I like the "Open in Zotero" option to open that item as a reader tab. There, instead of drag-dropping annotation into the "Cite tab", one could drag-drop it into the citation dialog (or add it via context menu option). What would be an additional benefit of duplicating `bubble-input` into the dedicated "Cite tab" here?

yexingsha commented:
> In this case, does that mean that there would be two active instances of bubble-input at the same time from where one can accept what to insert?

I was thinking that the original bubble-input would go away, so it feels like it's "moved into the main window", in a sense. 

> So what does a separate "Cite tab" in the item/context pane add VS just a context menu "Add to citation dialog" on the citation in the sidebar/itemTree/etc? [...] There, instead of drag-dropping annotation into the "Cite tab", one could drag-drop it into the citation dialog (or add it via context menu option).

Let's say we use the citation dialog approach. Once you use "Open in Zotero" to look for annotations in the reader, the citation dialog is no longer in front. So unless both windows are visible, there is no feasible way to drag-and-drop; and after you add via context menu option, you need to tab back to the dialog to accept. Since many people have trouble with tabbing, I thought a Cite tab would provide a smoother experience. But, if we can somehow bring the citation dialog to the front, then yes, the Cite tab doesn't add much.

That said, I do agree with @dstillman 's points here. Let's not worry about the Cite tab for now. Even if we can't solve the "needing to tab back from main window to citation dialog" issue, people can still navigate to main window to select annotations before they open the citation dialog, which is probably easier since they can use the taskbar/dock for that.

abaevbog commented:
That sounds good! Thank you for your feedback, @dstillman.

With this, let me try to summarize the direction of further updates below:
- keep the list mode hidden, at least for now
- selected annotations in the reader tab will count as selected annotations in the citation dialog, similar to annotations selected in itemTree. I assume that only applies to annotations selected in the current reader tab.
- without collapsible sections on top of itemTree for Selected/Open items, we go back to displaying item cards in the "Suggested items" section above the itemTree. To avoid having annotation cards mixed up with top-level item cards, we'll place them in separate sections. So there will be "Selected Annotations", "Selected Items", "Open Items", and "Cited Items" groups. Clicking the selected annotation will bubbleize it, clicking on a selected top-level item will select it in the itemTree.
- when an annotation bubble is clicked, and the popup with annotation preview from https://github.com/zotero/zotero/pull/5356#issuecomment-3051089977 will appear
- I added the sidebar preview of selected annotations to the right of `itemTree`. For now, I reused the existing `annotation-items-pane` custom XUL component. @adomasven, I remember during the initial steps of the citation dialog implementation, we talked about not using XUL as much as possible. Would that apply to this situation?
- I assume, we still want the "Open in Zotero" button to open an item from the itemTree. We could do it when a top-level item is activated (e.g. double-click or Enter) but it feels not an expected action for that. So would it be another button, similar to `+` and `-` that we have to annotations? What could that look like (cc @yexingsha)?

yexingsha commented:
I think "adding all annotations in one item" is probably not that common, so maybe we can replace the `+` of parent items and attachments with the "Open in Zotero" button. But then, would double clicking/enter still add all the annotations? Does that feel too inconsistent?

The button can look like this:

<img width="1000" height="700" alt="annotation-library" src="https://github.com/user-attachments/assets/ceab0276-4bbb-40cc-b1fa-492f9d812720" />

dstillman commented:
> maybe we can replace the + of parent items and attachments with the "Open in Zotero" button. But then, would double clicking/enter still add all the annotations? Does that feel too inconsistent?

That does feel a bit too inconsistent to me. I also worry that the icon isn't really clear — it looks kind of like it's going to launch you out of Zotero to the web.

And what would it do for an attachment? Take you to the library or the reader? And if the latter, what if you're not using the Zotero reader? Launch an external PDF reader? That doesn't seem particularly useful in this context.

I guess I'm not sure why we need this function at all. You're already in a library view. And if you want to select annotations in the reader, you don't really need to start with this dialog at all — you should just do it in Zotero and pull up the citation dialog. We have the button in the citation popup that takes you to the items list, but that's because you might want to edit the metadata, etc.

adomasven commented:
> I guess I'm not sure why we need this function at all. You're already in a library view. And if you want to select annotations in the reader, you don't really need to start with this dialog at all — you should just do it in Zotero and pull up the citation dialog. We have the button in the citation popup that takes you to the items list, but that's because you might want to edit the metadata, etc.

It's not obvious or intuitive to do this. Although we can think of other ways to make this workflow more discoverable. The placeholder text in the selected items area could be more descriptive, perhaps even some screenshots? In that case, maybe it shouldn't show selected items, only annotations?

dstillman commented:
I'm not sure why it wouldn't be obvious? It's now a very prominent feature in the other citation dialogs, so I don't know why people wouldn't understand that the same thing is possible in this dialog. I don't know that we've figured out the best way to actually display selected annotations, but the process itself would be consistent.

Or do you just mean the difference between selecting an item in the items list (which is a very clear selection state) vs. selecting an annotation in the reader (which is a less clear selection state)?

dstillman commented:
> But if you are in the middle of a writing process and are looking at the "Add annotation" dialog thinking "this detail was somewhere in this paper but where", "Go to Zotero" could be somewhat helpful?

Sure. I guess I'm just not sure how often going back to the document would be necessary, and whether people wouldn't usually know that in advance. You can search based on the text or comment in the dialog — that's one of the reasons to include the annotations in the dialog's items list. So you'd have to be searching or skimming the document for content that's not in the annotation you're trying to find.

Mostly I'm worried about potential confusion. We know people are bad at managing windows. We specifically put the citation dialog on top of Zotero (or, ideally, the word processor itself), and I think that if we encourage people to perform an action that sends the dialog behind the main window, a lot of people will have trouble getting back to the dialog. Yes, they can just get back to their word processor and hopefully click Add Annotation again and it will come to the front, but that's a bit messy.

If we really did think this was necessary, I guess I might actually favor the in-dialog reader after all:

1. It avoids the mess of windows.
2. We could actually do it for people who don't use the built-in reader normally.
3. It seems like the main thing you'd want here would specifically be the reader, so sending people back to the library view would just be a pointless intermediate step anyway.

How hard do we think the in-dialog reader would be?

dstillman commented:
> You can search based on the text or comment in the dialog — that's one of the reasons to include the annotations in the dialog's items list.

Also on this, in addition to searching the whole library for annotation text/comments, I'm assuming we'd have a filter bar in the right-hand pane as well, as we do in the reader sidebar, so you could click on the parent item, see all the annotations, and filter that way too.

abaevbog commented:
>   How hard do we think the in-dialog reader would be?

The reader itself, I expect, wouldn't be that bad. I briefly had a look at it during the initial discussion, and `attachmentBox` seems like a fairly good blueprint for a mini-reader, I think.

The sidebar with annotations may be slightly more challenging, since not all options from reader tabs/windows apply here, so I'm not sure if we would be able to properly reuse it. Let me look into this to get a better idea of what it may be like.

dstillman commented:
For now, in order to move forward, I would say we should not worry about this. We can investigate an in-dialog reader later.

But we do need to figure out how to handle selected/open in this dialog.

abaevbog commented:
> For now, in order to move forward, I would say we should not worry about this. We can investigate an in-dialog reader later.

That sounds good!

> But we do need to figure out how to handle selected/open in this dialog.

Besides the collapsible sections in itemTree, one of the suggestions was to display item cards the same way as it is done now in the citation dialog but have separate decks for selected annotations and selected top-level items. That way, item cards of different type are not mixed together. Clicking on the annotation card would bubbleize it, and clicking on the selected or open top level item could select it in itemTree and expand all child attachments?

abaevbog commented:
As I am working through these, I am realizing that a sizeable portion (more than a half) of changes in this PR is the [refactoring of list mode](https://github.com/zotero/zotero/pull/5356/commits/55edef206ca98781862b9f9852710a8ced58a1f0) to use a `VirtualizedTable` to be able to render an arbitrary number of annotations per https://github.com/zotero/zotero/pull/5356#issuecomment-3009946526. This would also apply to other dialogs for consistency. Now that we decided to not include list mode,  none of those changes are necessary. What should I do with those commits?

Rendering an arbitrary number of results is somewhat of an improvement for the citation dialog but probably very minor. Personally, I think it makes the code a bit cleaner. But at the same time, it's a complete redo of how the list mode worked, so some behaviors are a bit different. Should I stash those changes for now and make a separate PR later on? Or do I still keep them here?

All other changes related to what we discussed earlier (e.g. all list mode tweaks, pseudo collection, collapsible sections in itemTree), I am just going to discard.

dstillman commented:
"Keep Comments" → "Include Comments"?

That should be a dialog-wide setting (as originally shown by @yexingsha), no? I think that's going to mostly be a workflow preference that should remember its last setting. It should be enabled by default, in line with the current behavior when adding annotations to notes.

Is the above the default height of the dialog?

abaevbog commented:
> That should be a dialog-wide setting (as originally shown by yexingsha), no?

Oh, right, from the initial mockup. I'll move it into the pref at the bottom.

> It should be enabled by default, in line with the current behavior when adding annotations to notes.

Gotcha!

> Is the above the default height of the dialog? I think we'll want it a bit taller, to give more space to the items list and annotations pane on the right.

Yes, it starts with 500px, same as other dialogs. We should definitely bump it up to make more space and to offset the taller suggested items area.

abaevbog commented:
I believe now the previous notes should be addressed, with a few additional fixes that I added along the way:

- I increased the minimum initial height to 650px
- the "Include Annotations" preference is moved into the bottom of the dialog
- general keyboard navigation of the sidenav
- fixed some flickering with itemTree when an item is bubbleized
- fixed itemNode's appearing pre-selected on mousedown (sometimes)

How does the current behavior look?

A couple of things I am wondering about is:
- when no annotations are selected in the itemPane, currently, there is nothing. We could show some kind of helpful message there? But I can't think of what would be appropriate there
- just like in normal citation diallog mode, one can still add multiple annotations as bubbles. I can't think of a specific use case when one would want to do that but I also can't think of a good reason to disallow it. If all else is equal, I figured keeping it consistent with the citation dialog is probably best.

For reference, this is the dialog with the new initial height. Let me know what you think!

<img width="849" height="653" alt="Screenshot 2025-07-21 at 6 54 43 PM" src="https://github.com/user-attachments/assets/9acab4f5-05a7-4dbb-ad13-7c4d41ee54e2" />

yexingsha commented:
This is looking really good!

> when no annotations are selected in the itemPane, currently, there is nothing. We could show some kind of helpful message there? But I can't think of what would be appropriate there

Probably something like "Select an item, attachment, or annotation to view annotation details"?

> just like in normal citation dialog mode, one can still add multiple annotations as bubbles. I can't think of a specific use case when one would want to do that but I also can't think of a good reason to disallow it. If all else is equal, I figured keeping it consistent with the citation dialog is probably best.

I guess this depends on people's writing habits. Some people might prefer to grab every annotation that they might use all at once, and then organize them as they write. And consistency with the citation dialog is indeed good.

Issues I have encountered so far:

- In my mocks I had a side pane toggle button in the bottom right corner. After testing this, I found the side pane very useful (especially with the filter!) so there's probably little reason to toggle it off. We might want to increase the initial width of the dialog to 1000px instead.
- The filter input box should have 8px margin on both sides.
- The annotation cards (in preview and side pane) and the spacing in the side pane aren't following compact/comfortable settings.
- In preview, reduce its padding to 16px, and increase the spacing between sections to 16px.
- The quotes in the Selected Annotations cards don't need to be italicized; I think the differences in font size and color is enough to distinguish it.
- When clicking on Selected Items / Open documents, it only works when the item has annotations and exists in the current active collection. So we need to do some library/collection switching. As to items without annotations, we can either exclude them from the list altogether, or add a line to every card saying "X/No Annotation", which is what I did in the mocks.
- I'm still seeing unrelated child items (notes, attachments without annotations) in the item tree.

abaevbog commented:
Thanks for the great feedback!

> Probably something like "Select an item, attachment, or annotation to view annotation details"?

Yeah, that sounds good to me. I put that message in, and we can tweak it later it it needs to be a bit shorter.

> In my mocks I had a side pane toggle button in the bottom right corner. After testing this, I found the side pane very useful (especially with the filter!) so there's probably little reason to toggle it off. We might want to increase the initial width of the dialog to 1000px instead.

Right, I also felt like the side pane is the key component here. If you have more than just a few annotations, in most cases (I assume), one would click on the row in the table, look at the side pane to confirm that's the right one, and then add it. As far as the initial width goes, it is mainly handled by the OS. In most cases, it is the width that that dialog had last time when it was closed. I am a bit cautious of messing with the window handling - last time I touched the initial positioning of the dialog (which is treated the same way - new dialog appears where the old one was), it lead to unexpected side effects when multiple monitors were being used.

> The annotation cards (in preview and side pane) and the spacing in the side pane aren't following compact/comfortable settings.

Good catch! The annotation cards and the side pane should be taken care of. A few more things worth noting here. It's not only the density but also the font pref was not being respected. If one sets their font-size pref, it's probably reasonable to expect it to propagate to this dialog as well. So I tweaked the remaining fonts to be tied to `rem` vs a hardcoded pixel value.

As far as item cards in the suggested items section go, the density does not currently affect them. I'm not sure what exactly we can do there. Vertically there's barely enough space to fit the title of the parent with annotation text and comment. We could shrink them horizontally a bit - but I currently see no more than 3 words in each card, so shrinking it even more may be counter productive. Not sure - let me know what you think!

>When clicking on Selected Items / Open documents, it only works when the item has annotations and exists in the current active collection. So we need to do some library/collection switching.

Ah yes. I plugged in the same handling as everywhere else in the main window. If the current collection has the item - it'll be selected. Otherwise, it'll be selected in the parent Library.

> I'm still seeing unrelated child items (notes, attachments without annotations) in the item tree.

Right, we have an older open PR to hide context rows in itemTree during search https://github.com/zotero/zotero/pull/5315. When it is merged, it will either apply the config to all itemTree's (including this dialog) or we can force-hide context rows in this dialog as a special case.

yexingsha commented:
> As far as item cards in the suggested items section go, the density does not currently affect them. I'm not sure what exactly we can do there.

I don't think we need to give them density settings. The horizontal scroll means that any change in vertical spacing will have little overall effect, and the only way to meaningfully change information density is to change their width, which isn't really beneficial. They are some what analogous to tabs, which aren't affected by density either.

Also, the placeholder message in side pane should have 8px margins, and the annotation filter should have 8px bottom margin (visible when the content below is scrolled up).

abaevbog commented:
The new icons look great! I reverted the temp tweak to open up the "Add Annotations" dialog when "Insert Note" is pressed.

One small thing I noticed when trying out the .dotm for word on mac is that after inserting some text (via "Add annotation" or "Add note"), the last line for some reason ends up having a smaller font (11) than all other lines. And that happens if I insert a note using the current beta build as well. Not sure if it is related to this change or something else but I did not see it in word before trying out the new .dotm file.

For reference, this is with the current beta build:

https://github.com/user-attachments/assets/046698c1-4d92-40e7-a6e2-1c2f3d231fb4

adomasven commented:
> One small thing I noticed when trying out the .dotm for word on mac is that after inserting some text (via "Add annotation" or "Add note"), the last line for some reason ends up having a smaller font (11) than all other lines. And that happens if I insert a note using the current beta build as well. Not sure if it is related to this change or something else but I did not see it in word before trying out the new .dotm file.

I cannot reproduce this. Create a new note in Zotero and see if that still happens. If it does, that's a separate issue and unlikely to be related to the dotm. Create an issue and I'll take a look.

abaevbog commented:
Gotcha. Yeah, I still see it, so I added the linked issue with a more detailed recording using a new note.

adomasven commented on test/tests/citationDialogTest.js:
```diff
@@ -674,6 +677,175 @@ describe("Citation Dialog", function () {
+
+			bubble.click();
+			// give the popup time to open
+			await Zotero.Promise.delay(50);
```
This is failing locally for me on macOS. Please avoid this pattern and use `await waitForDOMEvent(popup, "popupshown");` or similar.
